File: python/src/veat.py
# VEAT: value → entity_id → set of attribute names
# Fast for: "which entities have this value?"

def index_add(index, datom):
    val = index.get(datom.value, {})
    existing = val.get(datom.entity_id, set())
    return {**index, datom.value: {**val, datom.entity_id: existing | {datom.attr_name}}}


def index_remove(index, datom):
    val = index.get(datom.value, {})
    new_set = val.get(datom.entity_id, set()) - {datom.attr_name}
    return {**index, datom.value: {**val, datom.entity_id: new_set}}


def index_get(index, value, entity_id=None):
    val = dict(index.get(value, {}))
    if entity_id is None:
        return val
    return val.get(entity_id, set())

File: clojure/src/circle_db/veat.clj
(ns circle-db.veat)
;; VEAT: value → entity-id → set of attribute names
;; Fast for: "which entities have this value?"

(defn index-add [index datom]
  (update-in index [(:value datom) (:entity-id datom)] (fnil conj #{}) (:attr-name datom)))

(defn index-remove [index datom]
  (update-in index [(:value datom) (:entity-id datom)] disj (:attr-name datom)))

(defn index-get
  ([index value]
   (get index value {}))
  ([index value entity-id]
   (get-in index [value entity-id] #{})))
